File: backend/src/api/v1/component_templates.rs
//! Component-template HTTP handlers (Component Editor design §3.5).
//!
//! Routers map `FromRow` models to `*Read`/`*Summary` DTOs via the service
//! layer; they never query the DB directly and never serialize a `FromRow`
//! struct. `{cid}` = component UUID; `{vnum}` = `version_number` i32.

use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    response::IntoResponse,
    routing::{get, post},
    Json, Router,
};
use serde::Deserialize;
use uuid::Uuid;

use crate::{
    error::AppResult,
    schemas::{
        component_template::{
            ComponentTemplateCreate, ComponentTemplateRead, ComponentTemplateSummary,
            ComponentTemplateUpdate, ComponentTemplateVersionRead, ResolvedComponentRead,
            VersionCreate, VersionUpdate,
        },
        pagination::{Page, PageParams},
    },
    services::component_template_service::{self, VersionSelector},
    state::AppState,
};

/// Build the component-template sub-router.
pub fn router() -> Router<AppState> {
    Router::new()
        .route("/component-templates", post(create).get(list))
        .route("/component-templates/by-slug/{slug}", get(get_by_slug))
        .route(
            "/component-templates/{cid}",
            get(get_one).patch(update).delete(delete),
        )
        .route(
            "/component-templates/{cid}/versions",
            post(create_version).get(list_versions),
        )
        .route(
            "/component-templates/{cid}/versions/{vnum}",
            get(get_version)
                .patch(update_version)
                .delete(delete_version),
        )
        .route(
            "/component-templates/{cid}/versions/{vnum}/make-default",
            post(make_default),
        )
        .route("/component-templates/{cid}/resolve", get(resolve))
}

/// Optional `?q` name filter for the list endpoint.
#[derive(Debug, Default, Deserialize)]
pub struct ComponentListQuery {
    /// Case-insensitive `name ILIKE '%q%'` filter.
    #[serde(default)]
    pub q: Option<String>,
}

/// `?version=default|N` query param for the resolve endpoint.
#[derive(Debug, Default, Deserialize)]
pub struct ResolveQuery {
    /// `default` (or absent) → current default; a positive integer → that version.
    #[serde(default)]
    pub version: Option<String>,
}

/// `POST /component-templates` — create a component + its v1.
#[utoipa::path(
    post,
    path = "/api/v1/component-templates",
    request_body = ComponentTemplateCreate,
    responses(
        (status = 201, description = "Created", body = ComponentTemplateRead),
        (status = 409, description = "Slug conflict", body = crate::error::ErrorEnvelope),
        (status = 422, description = "Validation error", body = crate::error::ErrorEnvelope)
    ),
    tag = "component_templates"
)]
pub async fn create(
    State(state): State<AppState>,
    Json(input): Json<ComponentTemplateCreate>,
) -> AppResult<impl IntoResponse> {
    let component = component_template_service::create(&state.pool, input).await?;
    Ok((StatusCode::CREATED, Json(component)))
}

/// `GET /component-templates?page&page_size&q` — paginated list.
#[utoipa::path(
    get,
    path = "/api/v1/component-templates",
    params(
        PageParams,
        ("q" = Option<String>, Query, description = "Case-insensitive name filter")
    ),
    responses((status = 200, description = "Component page", body = inline(Page<ComponentTemplateSummary>))),
    tag = "component_templates"
)]
pub async fn list(
    State(state): State<AppState>,
    Query(params): Query<PageParams>,
    Query(filter): Query<ComponentListQuery>,
) -> AppResult<Json<Page<ComponentTemplateSummary>>> {
    let page = component_template_service::list(&state.pool, &params, filter.q.as_deref()).await?;
    Ok(Json(page))
}

/// `GET /component-templates/{cid}` — fetch with version summaries.
#[utoipa::path(
    get,
    path = "/api/v1/component-templates/{cid}",
    params(("cid" = Uuid, Path, description = "Component id")),
    responses(
        (status = 200, description = "Component", body = ComponentTemplateRead),
        (status = 404, description = "Not found", body = crate::error::ErrorEnvelope)
    ),
    tag = "component_templates"
)]
pub async fn get_one(
    State(state): State<AppState>,
    Path(cid): Path<Uuid>,
) -> AppResult<Json<ComponentTemplateRead>> {
    let component = component_template_service::get_with_versions(&state.pool, cid).await?;
    Ok(Json(component))
}

/// `GET /component-templates/by-slug/{slug}` — fetch with version summaries by
/// slug. The static `by-slug` segment is registered before `{cid}` so it is not
/// shadowed by the UUID param route.
#[utoipa::path(
    get,
    path = "/api/v1/component-templates/by-slug/{slug}",
    params(("slug" = String, Path, description = "Component slug")),
    responses(
        (status = 200, description = "Component", body = ComponentTemplateRead),
        (status = 404, description = "Not found", body = crate::error::ErrorEnvelope)
    ),
    tag = "component_templates"
)]
pub async fn get_by_slug(
    State(state): State<AppState>,
    Path(slug): Path<String>,
) -> AppResult<Json<ComponentTemplateRead>> {
    let component =
        component_template_service::get_with_versions_by_slug(&state.pool, &slug).await?;
    Ok(Json(component))
}

/// `PATCH /component-templates/{cid}` — update metadata + default pointer.
#[utoipa::path(
    patch,
    path = "/api/v1/component-templates/{cid}",
    params(("cid" = Uuid, Path, description = "Component id")),
    request_body = ComponentTemplateUpdate,
    responses(
        (status = 200, description = "Updated", body = ComponentTemplateRead),
        (status = 404, description = "Not found", body = crate::error::ErrorEnvelope),
        (status = 409, description = "Slug conflict", body = crate::error::ErrorEnvelope),
        (status = 422, description = "Validation error", body = crate::error::ErrorEnvelope)
    ),
    tag = "component_templates"
)]
pub async fn update(
    State(state): State<AppState>,
    Path(cid): Path<Uuid>,
    Json(input): Json<ComponentTemplateUpdate>,
) -> AppResult<Json<ComponentTemplateRead>> {
    let component = component_template_service::update(&state.pool, cid, input).await?;
    Ok(Json(component))
}

/// `DELETE /component-templates/{cid}` — delete (cascades versions).
#[utoipa::path(
    delete,
    path = "/api/v1/component-templates/{cid}",
    params(("cid" = Uuid, Path, description = "Component id")),
    responses(
        (status = 204, description = "Deleted"),
        (status = 404, description = "Not found", body = crate::error::ErrorEnvelope)
    ),
    tag = "component_templates"
)]
pub async fn delete(
    State(state): State<AppState>,
    Path(cid): Path<Uuid>,
) -> AppResult<impl IntoResponse> {
    component_template_service::delete(&state.pool, cid).await?;
    Ok(StatusCode::NO_CONTENT)
}

/// `GET /component-templates/{cid}/versions` — list full version reads.
#[utoipa::path(
    get,
    path = "/api/v1/component-templates/{cid}/versions",
    params(("cid" = Uuid, Path, description = "Component id")),
    responses(
        (status = 200, description = "Versions", body = inline(Vec<ComponentTemplateVersionRead>)),
        (status = 404, description = "Not found", body = crate::error::ErrorEnvelope)
    ),
    tag = "component_templates"
)]
pub async fn list_versions(
    State(state): State<AppState>,
    Path(cid): Path<Uuid>,
) -> AppResult<Json<Vec<ComponentTemplateVersionRead>>> {
    let versions = component_template_service::list_versions(&state.pool, cid).await?;
    Ok(Json(versions))
}

/// `POST /component-templates/{cid}/versions` — create a new version.
#[utoipa::path(
    post,
    path = "/api/v1/component-templates/{cid}/versions",
    params(("cid" = Uuid, Path, description = "Component id")),
    request_body = VersionCreate,
    responses(
        (status = 201, description = "Created", body = ComponentTemplateVersionRead),
        (status = 404, description = "Not found", body = crate::error::ErrorEnvelope),
        (status = 422, description = "Validation error", body = crate::error::ErrorEnvelope)
    ),
    tag = "component_templates"
)]
pub async fn create_version(
    State(state): State<AppState>,
    Path(cid): Path<Uuid>,
    Json(input): Json<VersionCreate>,
) -> AppResult<impl IntoResponse> {
    let version = component_template_service::create_version(&state.pool, cid, input).await?;
    Ok((StatusCode::CREATED, Json(version)))
}

/// `GET /component-templates/{cid}/versions/{vnum}` — fetch one version.
#[utoipa::path(
    get,
    path = "/api/v1/component-templates/{cid}/versions/{vnum}",
    params(
        ("cid" = Uuid, Path, description = "Component id"),
        ("vnum" = i32, Path, description = "Version number")
    ),
    responses(
        (status = 200, description = "Version", body = ComponentTemplateVersionRead),
        (status = 404, description = "Not found", body = crate::error::ErrorEnvelope)
    ),
    tag = "component_templates"
)]
pub async fn get_version(
    State(state): State<AppState>,
    Path((cid, vnum)): Path<(Uuid, i32)>,
) -> AppResult<Json<ComponentTemplateVersionRead>> {
    let version = component_template_service::get_version(&state.pool, cid, vnum).await?;
    Ok(Json(version))
}

/// `PATCH /component-templates/{cid}/versions/{vnum}` — edit a version in place.
#[utoipa::path(
    patch,
    path = "/api/v1/component-templates/{cid}/versions/{vnum}",
    params(
        ("cid" = Uuid, Path, description = "Component id"),
        ("vnum" = i32, Path, description = "Version number")
    ),
    request_body = VersionUpdate,
    responses(
        (status = 200, description = "Updated", body = ComponentTemplateVersionRead),
        (status = 404, description = "Not found", body = crate::error::ErrorEnvelope),
        (status = 422, description = "Validation error", body = crate::error::ErrorEnvelope)
    ),
    tag = "component_templates"
)]
pub async fn update_version(
    State(state): State<AppState>,
    Path((cid, vnum)): Path<(Uuid, i32)>,
    Json(input): Json<VersionUpdate>,
) -> AppResult<Json<ComponentTemplateVersionRead>> {
    let version = component_template_service::update_version(&state.pool, cid, vnum, input).await?;
    Ok(Json(version))
}

/// `DELETE /component-templates/{cid}/versions/{vnum}` — delete a version.
#[utoipa::path(
    delete,
    path = "/api/v1/component-templates/{cid}/versions/{vnum}",
    params(
        ("cid" = Uuid, Path, description = "Component id"),
        ("vnum" = i32, Path, description = "Version number")
    ),
    responses(
        (status = 204, description = "Deleted"),
        (status = 404, description = "Not found", body = crate::error::ErrorEnvelope),
        (status = 409, description = "Last version protected", body = crate::error::ErrorEnvelope)
    ),
    tag = "component_templates"
)]
pub async fn delete_version(
    State(state): State<AppState>,
    Path((cid, vnum)): Path<(Uuid, i32)>,
) -> AppResult<impl IntoResponse> {
    component_template_service::delete_version(&state.pool, cid, vnum).await?;
    Ok(StatusCode::NO_CONTENT)
}

/// `POST /component-templates/{cid}/versions/{vnum}/make-default` — pin default.
#[utoipa::path(
    post,
    path = "/api/v1/component-templates/{cid}/versions/{vnum}/make-default",
    params(
        ("cid" = Uuid, Path, description = "Component id"),
        ("vnum" = i32, Path, description = "Version number")
    ),
    responses(
        (status = 200, description = "Default updated", body = ComponentTemplateRead),
        (status = 404, description = "Not found", body = crate::error::ErrorEnvelope)
    ),
    tag = "component_templates"
)]
pub async fn make_default(
    State(state): State<AppState>,
    Path((cid, vnum)): Path<(Uuid, i32)>,
) -> AppResult<Json<ComponentTemplateRead>> {
    let component = component_template_service::make_default(&state.pool, cid, vnum).await?;
    Ok(Json(component))
}

/// `GET /component-templates/{cid}/resolve?version=default|N` — proxy-facing
/// resolve. `Version(n)` falls back to the current default when missing.
#[utoipa::path(
    get,
    path = "/api/v1/component-templates/{cid}/resolve",
    params(
        ("cid" = Uuid, Path, description = "Component id"),
        ("version" = Option<String>, Query, description = "default | N")
    ),
    responses(
        (status = 200, description = "Resolved component", body = ResolvedComponentRead),
        (status = 404, description = "Not found", body = crate::error::ErrorEnvelope),
        (status = 422, description = "Bad version selector", body = crate::error::ErrorEnvelope)
    ),
    tag = "component_templates"
)]
pub async fn resolve(
    State(state): State<AppState>,
    Path(cid): Path<Uuid>,
    Query(query): Query<ResolveQuery>,
) -> AppResult<Json<ResolvedComponentRead>> {
    let selector = VersionSelector::parse(query.version.as_deref())?;
    let resolved = component_template_service::resolve(&state.pool, cid, selector).await?;
    Ok(Json(resolved))
}

File: frontend/src/lib/api/componentTemplates.ts
import {
  useMutation,
  useQuery,
  useQueryClient,
} from "@tanstack/react-query";
import { z } from "zod";

import { apiGet, apiSend, Page } from "@/lib/api/client";
import {
  ComponentTemplateCreate,
  ComponentTemplateRead,
  ComponentTemplateSummary,
  ComponentTemplateUpdate,
  ComponentTemplateVersionRead,
  ResolvedComponentRead,
  VersionCreate,
  VersionUpdate,
} from "@/lib/schemas/componentTemplates";

// Component-template API client + TanStack Query hooks (component-editor design
// §5.2). NOT `components.ts` — that file is the outcome sub-component client.
// All wire shapes mirror the locked backend DTOs (design §3.2); every response
// is zod-validated through the shared apiGet/apiSend helpers.

const BASE = "/api/v1/component-templates";

// ── Plain fetchers ───────────────────────────────────────────────────────────

export interface ListComponentTemplatesParams {
  page: number;
  page_size: number;
  q?: string;
}

export const listComponentTemplates = (
  p: ListComponentTemplatesParams = { page: 1, page_size: 20 },
) => {
  const params = new URLSearchParams({
    page: String(p.page),
    page_size: String(p.page_size),
  });
  if (p.q && p.q.trim() !== "") params.set("q", p.q.trim());
  return apiGet(`${BASE}?${params.toString()}`, Page(ComponentTemplateSummary));
};

export const getComponentTemplate = (cid: string) =>
  apiGet(`${BASE}/${cid}`, ComponentTemplateRead);

// Resolve a component by its route slug. Returns the same ComponentTemplateRead
// shape as getComponentTemplate; 404s when the slug is unknown.
export const getComponentTemplateBySlug = (slug: string) =>
  apiGet(`${BASE}/by-slug/${encodeURIComponent(slug)}`, ComponentTemplateRead);

export const createComponentTemplate = (b: ComponentTemplateCreate) =>
  apiSend("POST", BASE, ComponentTemplateRead, b);

export const updateComponentTemplate = (
  cid: string,
  b: ComponentTemplateUpdate,
) => apiSend("PATCH", `${BASE}/${cid}`, ComponentTemplateRead, b);

export const deleteComponentTemplate = (cid: string) =>
  apiSend("DELETE", `${BASE}/${cid}`, z.void());

export const listVersions = (cid: string) =>
  apiGet(`${BASE}/${cid}/versions`, z.array(ComponentTemplateVersionRead));

export const getVersion = (cid: string, vnum: number) =>
  apiGet(`${BASE}/${cid}/versions/${vnum}`, ComponentTemplateVersionRead);

export const createVersion = (cid: string, b: VersionCreate) =>
  apiSend("POST", `${BASE}/${cid}/versions`, ComponentTemplateVersionRead, b);

export const updateVersion = (cid: string, vnum: number, b: VersionUpdate) =>
  apiSend(
    "PATCH",
    `${BASE}/${cid}/versions/${vnum}`,
    ComponentTemplateVersionRead,
    b,
  );

export const deleteVersion = (cid: string, vnum: number) =>
  apiSend("DELETE", `${BASE}/${cid}/versions/${vnum}`, z.void());

export const makeDefaultVersion = (cid: string, vnum: number) =>
  apiSend(
    "POST",
    `${BASE}/${cid}/versions/${vnum}/make-default`,
    ComponentTemplateRead,
  );

// `version` is the string "default" or a positive version number. Proxy-facing,
// but reused by the editor preview for the resolved variable list.
export const resolveComponent = (cid: string, version: "default" | number) =>
  apiGet(
    `${BASE}/${cid}/resolve?version=${version}`,
    ResolvedComponentRead,
  );

// ── Query keys ───────────────────────────────────────────────────────────────
export const componentKeys = {
  all: ["component-templates"] as const,
  list: (params: ListComponentTemplatesParams) =>
    ["component-templates", "list", params] as const,
  detail: (cid: string) => ["component-templates", "detail", cid] as const,
  versions: (cid: string) =>
    ["component-templates", "versions", cid] as const,
  version: (cid: string, vnum: number) =>
    ["component-templates", "version", cid, vnum] as const,
};

// ── Hooks ────────────────────────────────────────────────────────────────────

export function useComponentTemplates(params: ListComponentTemplatesParams) {
  return useQuery({
    queryKey: componentKeys.list(params),
    queryFn: () => listComponentTemplates(params),
  });
}

export function useComponentTemplate(cid: string, enabled = true) {
  return useQuery({
    queryKey: componentKeys.detail(cid),
    queryFn: () => getComponentTemplate(cid),
    enabled,
  });
}

export function useComponentVersion(
  cid: string,
  vnum: number | null,
) {
  return useQuery({
    queryKey: vnum === null ? ["component-templates", "version", cid, "none"] : componentKeys.version(cid, vnum),
    queryFn: () => getVersion(cid, vnum as number),
    enabled: vnum !== null,
  });
}

export function useCreateComponentTemplate() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (b: ComponentTemplateCreate) => createComponentTemplate(b),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: componentKeys.all });
    },
  });
}

export function useUpdateComponentTemplate(cid: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (b: ComponentTemplateUpdate) => updateComponentTemplate(cid, b),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: componentKeys.all });
    },
  });
}

export function useDeleteComponentTemplate() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (cid: string) => deleteComponentTemplate(cid),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: componentKeys.all });
    },
  });
}

export function useCreateVersion(cid: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (b: VersionCreate) => createVersion(cid, b),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: componentKeys.detail(cid) });
      void qc.invalidateQueries({ queryKey: componentKeys.versions(cid) });
    },
  });
}

export function useUpdateVersion(cid: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({ vnum, body }: { vnum: number; body: VersionUpdate }) =>
      updateVersion(cid, vnum, body),
    onSuccess: (_data, { vnum }) => {
      void qc.invalidateQueries({ queryKey: componentKeys.detail(cid) });
      void qc.invalidateQueries({ queryKey: componentKeys.versions(cid) });
      void qc.invalidateQueries({ queryKey: componentKeys.version(cid, vnum) });
    },
  });
}

export function useDeleteVersion(cid: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (vnum: number) => deleteVersion(cid, vnum),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: componentKeys.detail(cid) });
      void qc.invalidateQueries({ queryKey: componentKeys.versions(cid) });
    },
  });
}

export function useMakeDefaultVersion(cid: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (vnum: number) => makeDefaultVersion(cid, vnum),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: componentKeys.detail(cid) });
      void qc.invalidateQueries({ queryKey: componentKeys.versions(cid) });
    },
  });
}
